[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none

kadeng opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #121808

There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @muchulee8 @aakhundov @ColinPeppler @amjames @desertfire @chauhang

kadeng committed:
[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none

There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge

jansel reviewed (approved):
Add a test?

kadeng commented:
> Add a test?

Makes sense, I will add one. At the moment, this change is blocked due to a CI failure. On first sight it looks like this change triggered something (or CI is broken)

lezcano commented:
the CI failure was just the CI being flaky, it looks like

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng commented on test/inductor/test_torchinductor_dynamic_shapes.py:
```diff
@@ -609,8 +610,92 @@ def func(x, fn, a):
 
-instantiate_device_type_tests(TestInductorDynamic, globals())
+# @TODO: Uncomment again
+# instantiate_device_type_tests(TestInductorDynamic, globals())
```
> Add a test?

@jansel I have added one by now. I think it uncovered another preexisting issue that you might want to triage. If I uncomment this line, I get the following exception:

```
  File "/home/klondenberg/github/pytorch/pytorch/torch/_dynamo/guards.py", line 705, in SHAPE_ENV
    guards = output_graph.shape_env.produce_guards(
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/klondenberg/github/pytorch/pytorch/torch/fx/experimental/symbolic_shapes.py", line 3308, in produce_guards
    raise ConstraintViolationError(
torch.fx.experimental.symbolic_shapes.ConstraintViolationError: Constraints violated (L['x'].size()[0])! For more information, run with TORCH_LOGS="+dynamic".
  - Not all values of RelaxedUnspecConstraint(L['x'].size()[0]) are valid because L['x'].size()[0] was inferred to be a constant (5).
```

I think it's not a problem with this code. If I understand it all correctly, this Exception is triggered because a value that's marked to be dynamic is specialized to a static value in fn_3. But should this not just specialize and trigger recompilation if the guard is violated with a different input?

lezcano replied:
This is expected, as you are asking for the dymension to be dynamic, and then it's specialised to be static. These `TestInductorDynamic` just runs the tests with dynamic shapes, but in this test you are handling the dynamic shapes yourself, so no need to instantiate them with `TestInductorDynamic`

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano reviewed (commented):
This test is quite complex. Isn't it possible to test this function similar to what we do at
https://github.com/pytorch/pytorch/blob/4aaf25bc3813bfcf466d8758977f7433c27b9049/test/inductor/test_torchinductor.py#L8442-L8459

kadeng commented:
> This test is quite complex. Isn't it possible to test this function similar to what we do at
> 
> https://github.com/pytorch/pytorch/blob/4aaf25bc3813bfcf466d8758977f7433c27b9049/test/inductor/test_torchinductor.py#L8442-L8459

thanks, that looks right and will likely execute much faster.  

I arrived at the test you see here after writing it in different ways - also looking at how tests were set up in test_torchinductor_dynamic_shapes.py - but I was often hitting the ConstraintViolationError I mentioned above when constraining it to a static int in one way or another. 

So, to exclude the possibility that I was setting up the shape env in the wrong way for testing, I used the one that is actually created by the full logic of torch.compile with inductor backend.

lezcano commented on test/inductor/test_torchinductor_dynamic_shapes.py:
```diff
@@ -609,8 +610,92 @@ def func(x, fn, a):
+        torch._dynamo.reset()
+
+        _x = torch.randn([5, 3, 3])
+        torch._dynamo.mark_dynamic(_x, 0)
```
Just use `maybe_mark_dynamic` here.

lezcano commented on test/inductor/test_torchinductor_dynamic_shapes.py:
```diff
@@ -609,8 +610,92 @@ def func(x, fn, a):
+                return x.cos()
+            if x.shape[0] < 5:
+                return x * 2
+            # x.shape[0] == 5 at this point
```
Funnily enough, this should also error out when using `mark_dynamic`. Mind filing an issue? cc @ezyang

lezcano replied:
cc @avikchaudhuri

kadeng replied:
Filed https://github.com/pytorch/pytorch/issues/122307

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

kadeng committed:
Update on "[Inductor] Fix for WrapperCodeGen.statically_known_int_or_none"


There's obviously a small typo in WrapperCodeGen.statically_known_int_or_none,
where the return value of a call to V.graph._shape_env._maybe_evaluate_static
is being discarded.

This fix changes that to work how it was likely intended to.

Test Plan:
CI

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 muchulee8 aakhundov ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

lezcano reviewed (approved):
Thank you!

kadeng committed:
Rebasing, trying to get a green CI state

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
